tools/loader/ImageClassifier.cpp

```diff
@@ -430,25 +440,50 @@ int main(int argc, char **argv) {
   llvm::outs() << "Model: " << loader.getFunction()->getName() << "\n";
 
   int numErrors = 0;
+
+  // Image tensors loaded up to be run at once for benchmark mode.
+  std::vector<std::unique_ptr<Tensor>> preLoadedImages;
+  std::vector<std::unique_ptr<ExecutionContext>> contexts;
+
   while ((streamInputFilenamesMode &&
           getNextImageFilenames(&inputImageBatchFilenames)) ||
          (miniBatchMode &&
           getNextMiniBatch(inputImageBatchFilenames, inputImageFilenames,
                            minibatchIndex, miniBatch)) ||
          isFirstRun) {
-    // Load and process the image data into the inputImageData Tensor.
-    loadImagesAndPreprocess(inputImageBatchFilenames, &inputImageData,
-                            imageNormMode, imageChannelOrder, imageLayout);
-
+    auto inputImageData = llvm::make_unique<Tensor>();
+    if (iterationsOpt) {
+      // Load one image so we know dimensions and can create the appropriate
+      // batch size.
+      streamInputFilenamesMode = false;
+      miniBatchMode = false;
+      std::vector<std::string> firstImage;
+      firstImage.push_back(inputImageBatchFilenames[0]);
+      loadImagesAndPreprocess(firstImage, inputImageData.get(), imageNormMode,
+                              imageChannelOrder, imageLayout);
+      ShapeVector imageSize(inputImageData->getType().dims().begin(),
+                            inputImageData->getType().dims().end());
+      if (miniBatch) {
+        imageSize[0] = miniBatch;
+      } else {
+        imageSize[0] = iterationsOpt;
```

Comment:
This doesn't seem right, I don't think I should need to allocate `iterations` times as much memory to run a benchmark `iterations` times.  Coupled with the requests below I think we might be using something like `iterations`^2  as much memory as needed...

Reply:
In the else branch we will only allocate iterationsOpt worth of memory because the enclosing loop will only iterate once.

Reply:
I'm feeling a bit confused by the overlap between the number of iterations to run (& total number of images?) and the batchSize. It would be good to have the ability to run any number of iterations over any number of images with any batch size.

Reply:
mini-batch is the number of images per request, iterations is total images.
so if mini-batch is 10 and total images is 50 you will run 5 requests.